File: vigenere/cpp/encrypt.cpp
#include <vector>
#include <iostream>
#include <string>
#include <ctype.h>

namespace Vigenere {
	std::string encrypt(const std::string& str, const std::string& key) {
		std::string s = str;
		
		std::vector<int> k(key.length());

		int key_len = key.length();

		int i;
		for (i = 0; i < key_len; i++) {
			k[i] = tolower(key[i]) - 'a';
		}

		i = 0;
		for (auto& c : s) {
			if (islower(c)) {
				c = (c - 'a' + k[i++ % key_len]) % 26 + 'a';
			} else if (isupper(c)) {
				c = (c - 'A' + k[i++ % key_len]) % 26 + 'A';
			}
		}

		return s;
	}

	std::string decrypt(const std::string& str, const std::string& key) {
		std::string k = key;

		for (auto& c : k) {
			c = tolower(c) - 'a' - 1;
			c = 'z' - c;
		}

		return encrypt(str,k);
	}
};

File: vigenere/ruby/encrypt.rb
module Vigenere
	def Vigenere.encrypt(str, key)
		k = Array.new(key.length, 0)

		for i in 0...key.length
			k[i] = key[i].downcase[0].ord - 'a'.ord
		end

		i = 0
		s = []
		str.each_byte do |c|
			if c.chr =~ /[[:alpha:]]/
				if c.chr == c.chr.downcase
					s << ((c - 'a'.ord + k[i % key.length]) % 26 + 'a'.ord).chr
				elsif c.chr == c.chr.upcase
					s << ((c - 'A'.ord + k[i % key.length]) % 26 + 'A'.ord).chr
				end
				i+=1
			else
				s << c.chr
			end
		end

		return s.join('')
	end

	def Vigenere.decrypt(str,key)
		k = []
		key.each_byte do |c|
			k << ('z'.ord - (c.chr.downcase.ord - 'a'.ord - 1)).chr
		end

		return encrypt(str,k.join(''))
	end
end